### PR TITLE
observability in ml, ready for review

### PR DESCRIPTION
**Pull request by katyamust**
The article is ready for review.

### CODE REVIEW

File: observability/ml-observability.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# Observability in Machine Learning 
+
```
Comment (NarmathaBala) [inline observability/ml-observability.md:2]: @katyamust - Will be good to have some introduction to why observability is important (or what is unique in ML for observability)

File: observability/ml-observability.md [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+In re-tranining scenario model behaviour may degrade due to changes in data.
+Here is how to observe and report model performance in production.
+
+    Model deployment:
+    Deployed as dependency (pickle file)  
+    Deployed as service
+
+
+## Data versioning 
+
```
Comment (NarmathaBala) [inline observability/ml-observability.md:32]: Any recommended practices?

File: observability/ml-observability.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+
```
Comment (NarmathaBala) [inline observability/ml-observability.md:34]: @katyamust - looks great for an initial version

File: observability/ml-observability.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+## Model performance over time: data drift  
+
+In re-tranining scenario model behaviour may degrade due to changes in data.
+Here is how to observe and report model performance in production.
+
+    Model deployment:
+    Deployed as dependency (pickle file)  
+    Deployed as service
+
+## Data versioning
```
[the remark this answers is not in the document]
Comment (katyamust) [reply observability/ml-observability.md:41]: This was draft PR not ready for review. I have now updated the text and published PR.

File: observability/ml-observability.md (hunk 2 above)
Comment (katyamust) [reply observability/ml-observability.md:19]: Yes, to observe experimentation recommended practice is to use on of following tools: MLFLow for Databricks or Azure ML.

File: observability/ml-observability.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
+# Observability in Machine Learning  
+
+Development process of software system with machine learning component is more complex.
```
[the remark this answers is not in the document]
Comment (katyamust) [reply observability/ml-observability.md]: That is right, that was draft PR, not ready for review. Now published.

File: observability/ml-observability.md [hunk 6]
```diff
@@ -0,0 +1,44 @@
+# Observability in Machine Learning  
+
+Development process of software system with machine learning component is more complex
```
Comment (TessFerrandez) [inline observability/ml-observability.md:3]: Nit: language details

The Development process of a software system with machine learning components is more complex than traditional software. We need to monitor changes and variations in both code, the model and the data.

File: observability/ml-observability.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+than traditional software as such systems change in three directions: the code, the model and the data.
+We can distinguish 2 stages of such system lifespan: experimentation and production.
```
Comment (TessFerrandez) [inline observability/ml-observability.md:5]: Nit: language details

The two stages of these types of systems, experimentation and production, require different approaches to observability as discussed below:

File: observability/ml-observability.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+Both stages require specific approach to observability that we will discuss below:
+
+## Model experimentation and tuning
+
+Experimentation is a process of finding suitable machine learning model and its parameters via training and evaluating such models with one or more data sets.
```
Comment (TessFerrandez) [inline observability/ml-observability.md:10]: Nit: language details

Experimentation is the process of finding a suitable machine learning model and its parameters via training and evaluating the model with one or more datasets.

When developing and tuning machine learning models, the data scientists are interested in observing and comparing selected performance metrics for different parameters. They also need a reliable way to reproduce a training process, such that a given dataset and given parameters produces the same models.

File: observability/ml-observability.md [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+MLFlow is open source framework that can be hosted on Azure Databricks. You can use MLFlow SDK to log your evaluation metrics or any parameter you would like and track it at experimentation board in Azure Databricks.
+Source code and dataset version are also saved with log snapshot to provide reproducibility.
+
+[Azure Machine Learning Service](https://ml.azure.com/)
+Azure Machine Learning service provides SDK for Python, R and C#  to log your evaluation metrics to AML Experiment. Experiments are viewed in AML dashboard. Reproducibility is achived by storing code or notebook
+snapshot together with viewed metric. You can create versioned Datasets within Azure Machine Learning service.
+
+## Model in production
+
+Trained model can be deployed to production as container. Azure Machine Learning service provides SDK to deploy model as Azure Container Instance and publishes REST endpoint. You can monitor it using microservice observability methods.
```
Comment (TessFerrandez) [inline observability/ml-observability.md:26]: Nit: 
The trained model...
The Azure Machine Learning service...

Comment (TessFerrandez) [inline observability/ml-observability.md:26]: Is there any documentation we can point to for microservice observability methods that are relevant?

File: observability/ml-observability.md [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+MLFLow is alternative way to deploy ML model as a service.
```
Comment (TessFerrandez) [inline observability/ml-observability.md:27]: Nit: is an alternative way

File: observability/ml-observability.md [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+## Training and re-training
+
+To automatically retrain the model you can use AML Pipelines or Azure Databricks.
+When re-training with AML Pipelines you can monitor full information of each run including
```
Comment (TessFerrandez) [inline observability/ml-observability.md:32]: Nit: language details

To automatically re-train the model you can use AML Pipelines and Azure Databricks.
When re-training with AML Pipelines you can monitor information of each run, including the output, logs, and various metrics in the Azure portal experiment dashboard, or manually extract it using the AML SDK

File: observability/ml-observability.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+output,log and metric in the Azure portal experiment dashboard or manually extract it using SDK.
+
+## Model performance over time: data drift  
+
+We re-train machine learning models to improve its performance and make models better aligned with data changing over time. However, in some cases model performance may degrade. This may happen in case data change dramatically and do not expose pattern we observed during model development any more. This effect is called data drift.
```
Comment (TessFerrandez) [inline observability/ml-observability.md:37]: Nit: improve their performance

Nit: do not exhibit the patterns we observed during ...

File: observability/ml-observability.md [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+AML has preview feature to observe and report data drift.
+This [article](https://docs.microsoft.com/en-us/azure/machine-learning/how-to-monitor-datasets) describes it in detail.
+
+## Data versioning  
+
+It is recommended practice to add version to all datasets. Create Azure Ml Dataset with version for AML environment for this purpose.
```
Comment (TessFerrandez) [inline observability/ml-observability.md:43]: Nit:
You can create a versioned Azure ML Dataset for this purpose, or manually version it if using other systems.

File: observability/ml-observability.md (hunk 7 above)
Comment (NarmathaBala) [inline observability/ml-observability.md:5]: Suggest to be spelled out as 'two' for consistency

File: observability/ml-observability.md [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+with various parameters and models. Team is interested to observe and compare chosen performance metrics and be able to reproduce machine learning model, i.e. re-run training process with given dataset and parameters and produce the same model.
+
+The following tools are available to monitor model development process:
+
+[MLFlow for Databricks](https://docs.microsoft.com/en-us/azure/databricks/applications/mlflow/)
+MLFlow is open source framework that can be hosted on Azure Databricks. You can use MLFlow SDK to log your evaluation metrics or any parameter you would like and track it at experimentation board in Azure Databricks.
+Source code and dataset version are also saved with log snapshot to provide reproducibility.
+
+[Azure Machine Learning Service](https://ml.azure.com/)
+Azure Machine Learning service provides SDK for Python, R and C#  to log your evaluation metrics to AML Experiment. Experiments are viewed in AML dashboard. Reproducibility is achived by storing code or notebook
```
Comment (NarmathaBala) [inline observability/ml-observability.md:21]: Suggest to have Azure Machine Learning service (AML) so that AML can be referenced later

File: observability/ml-observability.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+## Model performance over time: data drift  
+
+We re-train machine learning models to improve its performance and make models better aligned with data changing over time. However, in some cases model performance may degrade. This may happen in case data change dramatically and do not expose pattern we observed during model development any more. This effect is called data drift.
+AML has preview feature to observe and report data drift.
+This [article](https://docs.microsoft.com/en-us/azure/machine-learning/how-to-monitor-datasets) describes it in detail.
+
+## Data versioning  
+
+It is recommended practice to add version to all datasets. Create Azure Ml Dataset with version for AML environment for this purpose.
+Use manual versioning in other systems.
```
Comment (NarmathaBala) [inline observability/ml-observability.md:44]: @katyamust - Is there a list of checklist that the lead can work with on ML. Certainly not as this PR but will be a great add as a separate PR

File: observability/ml-observability.md (hunk 13 above)
Comment (NarmathaBala) [inline observability/ml-observability.md:43]: Great content, thanks for your contribution

File: observability/ml-observability.md (hunk 8 above)
Comment (katyamust) [reply observability/ml-observability.md:10]: Nice wording! Thank you!

File: observability/ml-observability.md (hunk 9 above)
Comment (katyamust) [reply observability/ml-observability.md:26]: Added reference to Recipes section

File: observability/ml-observability.md (hunk 15 above)
Comment (katyamust) [reply observability/ml-observability.md]: @NarmathaBala I am not aware of such checklist, but happy to add one or contribute if there is already ongoing effort.
